be/src/http/action/tablet_migration_action.cpp

```diff
@@ -0,0 +1,258 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "http/action/tablet_migration_action.h"
+
+#include <string>
+
+#include "gutil/strings/substitute.h"
+#include "http/http_channel.h"
+#include "http/http_headers.h"
+#include "http/http_request.h"
+#include "http/http_status.h"
+#include "olap/storage_engine.h"
+#include "olap/tablet_manager.h"
+#include "olap/task/engine_storage_migration_task.h"
+#include "util/json_util.h"
+
+namespace doris {
+
+const static std::string HEADER_JSON = "application/json";
+
+TabletMigrationAction::TabletMigrationAction() {
+    _init_migration_action();
+}
+
+void TabletMigrationAction::_init_migration_action() {
+    int32_t max_thread_num = config::max_tablet_migration_threads;
+    int32_t min_thread_num = config::min_tablet_migration_threads;
+    ThreadPoolBuilder("MigrationTaskThreadPool")
+            .set_min_threads(min_thread_num)
+            .set_max_threads(max_thread_num)
+            .build(&_migration_thread_pool);
+}
+
+void TabletMigrationAction::handle(HttpRequest* req) {
+    int64_t tablet_id = 0;
+    int32_t schema_hash = 0;
+    string dest_disk = "";
+    string goal = "";
+    Status status = _check_param(req, tablet_id, schema_hash, dest_disk, goal);
+    if (status.ok()) {
+        if (goal == "run") {
+            MigrationTask current_task(tablet_id, schema_hash, dest_disk);
+            TabletSharedPtr tablet;
+            DataDir* dest_store;
+            Status status =
+                    _check_migrate_request(tablet_id, schema_hash, dest_disk, tablet, &dest_store);
+            if (status.ok()) {
+                do {
+                    {
+                        std::unique_lock<std::mutex> lock(_migration_status_mutex);
+                        std::map<MigrationTask, std::string>::iterator it_task =
+                                _migration_tasks.begin();
+                        for (; it_task != _migration_tasks.end(); it_task++) {
+                            if (tablet_id == (it_task->first)._tablet_id &&
+                                schema_hash == (it_task->first)._schema_hash) {
+                                status = Status::AlreadyExist(strings::Substitute(
+                                        "There is a migration task for this tablet already exists. "
+                                        "dest_disk is $0 .",
+                                        (it_task->first)._dest_disk));
+                                break;
+                            }
+                        }
+                        if (it_task != _migration_tasks.end()) {
+                            break;
+                        }
+                    }
+                    {
+                        std::unique_lock<std::mutex> lock(_migration_status_mutex);
```

Comment:
Looks like you can move this block to the above code block, to avoid lock `_migration_status_mutex ` twice?

Reply:
And the `"submitted"` status seems not necessary? It will be changed to "running" very soon. And it it submit failed, it will be removed from _migration_tasks.

Reply:
> Looks like you can move this block to the above code block, to avoid lock `_migration_status_mutex ` twice?

OK.

Reply:
> And the `"submitted"` status seems not necessary? It will be changed to "running" very soon. And it it submit failed, it will be removed from _migration_tasks.

When doing tablet rebalance between disks, there may be many migration tasks being submitted and waiting in the queue. So `"submitted"` status may be useful.